src/flint/flint_base/flint_base.pyx

```diff
@@ -114,16 +119,117 @@ cdef class flint_poly(flint_elem):
                 v = - fac[0]
                 roots.append((v, m))
         return roots
-    
+
     def complex_roots(self):
         raise AttributeError("Complex roots are not supported for this polynomial")
 
 
+cdef class flint_mpoly_context(flint_elem):
+    """
+    Base class for multivariate ring contexts
+    """
+
+    _ctx_cache = None
+
+    def __cinit__(self):
+        self._init = False
+
+    def __init__(self, int nvars, names):
+        if nvars < 0:
+            raise ValueError("cannot have a negative amount of variables")
+        elif len(names) != nvars:
+            raise ValueError("number of variables must match lens of variable names")
+        self.py_names = tuple(name.encode("ascii") if not isinstance(name, bytes) else name for name in names)
+        self.c_names = <const char**> libc.stdlib.malloc(nvars * sizeof(const char *))
+        self._init = True
+        for i in range(nvars):
+            self.c_names[i] = self.py_names[i]
+
+    def __dealloc__(self):
+        if self._init:
+            libc.stdlib.free(self.c_names)
+        self._init = False
```

Comment:
It should be fine to call `free` on a NULL pointer and I think that per default `self.c_names` will be a NULL pointer (i.e. Cython does this for us regardless of `__cinit__`). That should mean that we don't need `_init` or `__cinit__` and we can just call `free` unconditionally here.

Reply:
Agreed, I'm not sold on the whole `_init` var anyway. They're only used in a couple places, and I believe, only to prevent a freeing a null pointer using flint functions, which cannot handle null pointers and to prevent a double free. Cython doesn't guarantee that the `__dealloc__` function is only called once.
```
λ rg -g "*.pyx" "\._init[^\w]"
src/flint/flint_base/flint_base.pyx
135:        self._init = False
144:        self._init = True
149:        if self._init:
151:        self._init = False

src/flint/types/fmpz_mpoly_q.pyx
11:    var._init = True
18:    var._init = True
28:        self._init = False
31:        if self._init:
33:            self._init = False
44:                self._init = True

src/flint/types/fmpz_mpoly.pyx
146:        self._init = False
149:        if self._init:
151:            self._init = False

src/flint/types/dirichlet.pyx
26:        self._init = 0
29:        if self._init != 0:
35:        self._init = 1

src/flint/types/fmpq_mpoly.pyx
155:        self._init = False
158:        if self._init:
160:            self._init = False

```